### PR TITLE
[BEAM-1815] Avoid shuffling twice in GABW.

### PR DESCRIPTION
**Pull request by amitsela**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by amitsela** [comment]
R: @aviemzur

**Comment by amitsela** [comment]
Spark Run RunnableOnService

**Comment by amitsela** [comment]
Run Spark RunnableOnService

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupCombineFunctions.java [hunk 1]
```diff
@@ -49,18 +53,48 @@
...
-        .mapToPair(CoderHelpers.fromByteFunctionIterable(keyCoder, wvCoder))
-        .map(TranslationUtils.<K, Iterable<WindowedValue<V>>>fromPairFunction())
-        .map(WindowingHelpers.<KV<K, Iterable<WindowedValue<V>>>>windowFunction());
+    JavaPairRDD<ByteArray, byte[]> pairRDD =
+        rdd
+            .map(new ReifyTimestampsAndWindowsFunction<K, V>())
+            .map(WindowingHelpers.<KV<K, WindowedValue<V>>>unwindowFunction())
+            .mapToPair(TranslationUtils.<K, WindowedValue<V>>toPairFunction())
+            .mapToPair(CoderHelpers.toByteFunction(keyCoder, wvCoder));
+    // Use the RDD partitioner, if exists.
```
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupCombineFunctions.java]: Is this comment correct? Looks like we're creating a `HashPartitioner` not using the RDD partitioner.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/CoderHelpers.java [hunk 2]
```diff
@@ -193,18 +195,40 @@ public T call(byte[] bytes) throws Exception {
...
+      public Iterable<Tuple2<K, Iterable<V>>> call(
+          Iterator<Tuple2<ByteArray, Iterable<byte[]>>> itr) {
+        final Iterator<Tuple2<K, Iterable<V>>> outputItr =
+            Iterators.transform(
+                itr,
+                new com.google.common.base.Function<
+                    Tuple2<ByteArray, Iterable<byte[]>>, Tuple2<K, Iterable<V>>>() {
+                      @Override
+                      public Tuple2<K, Iterable<V>> apply(Tuple2<ByteArray, Iterable<byte[]>> t2) {
+                        K k = fromByteArray(t2._1().getValue(), keyCoder);
```
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/CoderHelpers.java]: Indentation seems off here.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/CoderHelpers.java [hunk 3]
```diff
@@ -193,18 +195,40 @@ public T call(byte[] bytes) throws Exception {
...
-      public Tuple2<K, Iterable<V>> call(Tuple2<ByteArray, Iterable<byte[]>> tuple) {
-        return new Tuple2<>(fromByteArray(tuple._1().getValue(), keyCoder),
-          Iterables.transform(tuple._2(), new com.google.common.base.Function<byte[], V>() {
-            @Override
-            public V apply(byte[] bytes) {
-              return fromByteArray(bytes, valueCoder);
-            }
-          }));
+      public Iterable<Tuple2<K, Iterable<V>>> call(
+          Iterator<Tuple2<ByteArray, Iterable<byte[]>>> itr) {
```
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/CoderHelpers.java]: Indentation seems off here.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupCombineFunctions.java (hunk 1 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupCombineFunctions.java]: Leftovers.. I'll fix that.

**Comment by amitsela** [comment]
@aviemzur PTAL

**Comment by amitsela** [comment]
`org.apache.beam.examples.WindowedWordCountIT.testWindowedWordCountInStreaming` failure is Flink, and `org.apache.beam.runners.apex.ApexRunnerTest.testConfigProperties` is Apex so disregard the failures.

**Comment by amitsela** [comment]
Run Spark RunnableOnService

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TranslationUtils.java [hunk 4]
```diff
@@ -138,7 +140,33 @@ public T2 call(Tuple2<T1, T2> v1) throws Exception {
     };
   }
 
-  /**  A pair to {@link KV} function . */
+  /** {@link KV} to pair flatmap function. */
+  public static <K, V> PairFlatMapFunction<Iterator<KV<K, V>>, K, V> toPairFlatMapFunction() {
```
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TranslationUtils.java]: Rename suggestion: `flatMapKVToPair` / `kvToPairFlatMapFunction`?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TranslationUtils.java [hunk 5]
```diff
@@ -148,22 +176,48 @@ public T2 call(Tuple2<T1, T2> v1) throws Exception {
     };
   }
 
-  /** Extract key from a {@link WindowedValue} {@link KV} into a pair. */
-  public static <K, V> PairFunction<WindowedValue<KV<K, V>>, K, WindowedValue<KV<K, V>>>
-      toPairByKeyInWindowedValue() {
-        return new PairFunction<WindowedValue<KV<K, V>>, K, WindowedValue<KV<K, V>>>() {
+  /** A pair to {@link KV} flatmap function . */
+  static <K, V> FlatMapFunction<Iterator<Tuple2<K, V>>, KV<K, V>> fromPairFlatMapFunction() {
```
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TranslationUtils.java]: Rename suggestion: `flatMapPairToKV` / `pairToKVFlatMapFunction`?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TranslationUtils.java [hunk 6]
```diff
@@ -270,9 +322,73 @@ public static void rejectStateAndTimers(DoFn<?, ?> doFn) {
...
     return new VoidFunction<T>() {
-      @Override public void call(T t) throws Exception {
+      @Override
+      public void call(T t) throws Exception {
         // Empty implementation.
       }
     };
   }
+
+  public static <T, K, V> PairFlatMapFunction<Iterator<T>, K, V> pairFunctionToPairFlatMapFunction(
```
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TranslationUtils.java]: This is a v. useful public function, we should probably have a Javadoc for it.
Same comment for `functionToFlatMapFunction`

**Comment by amitsela** [comment]
Run Spark RunnableOnService

**Pull request closed without merging**